File: frontend/src/services/order.ts
import { ApiError } from './auth'
import type {
  CreateOrderPayload,
  Order,
  OrderItem,
  OrderItemStoreMini,
  OrderStatus,
} from '../types/order'

const API_BASE = '/api/v1'

function mapStore(raw: Record<string, unknown> | null | undefined): OrderItemStoreMini | null {
  if (!raw) return null
  return {
    id: raw.id as string,
    name: raw.name as string,
    slug: (raw.slug as string | null) ?? null,
    avatarUrl: (raw.avatar_url as string | null) ?? null,
  }
}

function mapItem(raw: Record<string, unknown>): OrderItem {
  const image = (raw.product_image as string | null) ?? null
  const images = image
    ? image.split('|').map(s => s.trim()).filter(Boolean)
    : []
  return {
    id: raw.id as string,
    orderId: raw.order_id as string,
    productId: raw.product_id as string,
    productName: raw.product_name as string,
    quantity: raw.quantity as number,
    price: raw.price as number,
    notes: (raw.notes as string | null) ?? null,
    productImage: image,
    productImages: images,
    store: mapStore(raw.store as Record<string, unknown> | null),
  }
}

function mapOrder(raw: Record<string, unknown>): Order {
  return {
    id: raw.id as string,
    userId: raw.user_id as string,
    place: raw.place as string,
    phone: raw.phone as string,
    clientName: raw.client_name as string,
    totalAmount: raw.total_amount as number,
    status: raw.status as OrderStatus,
    note: (raw.note as string | null) ?? null,
    createdAt: (raw.created_at as string | null) ?? null,
    orderItems: (raw.order_items as Record<string, unknown>[]).map(mapItem),
  }
}

export async function createOrder(
  token: string,
  payload: CreateOrderPayload,
): Promise<Order> {
  const res = await fetch(`${API_BASE}/orders/`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      Authorization: `Bearer ${token}`,
    },
    body: JSON.stringify({
      place: payload.place,
      phone: payload.phone,
      client_name: payload.clientName,
      total_amount: payload.totalAmount,
      note: payload.note ?? null,
      items: payload.items.map(i => ({
        product_id: i.productId,
        product_name: i.productName,
        quantity: i.quantity,
        price: i.price,
        notes: i.notes ?? null,
      })),
      cart_item_ids: payload.cartItemIds ?? null,
    }),
  })
  if (!res.ok) {
    const data = await res.json().catch(() => ({}))
    throw new ApiError(res.status, data.detail ?? 'Create order failed')
  }
  return mapOrder(await res.json())
}

export async function listOrders(
  token: string,
  status?: OrderStatus,
): Promise<Order[]> {
  const url = status
    ? `${API_BASE}/orders/?status=${encodeURIComponent(status)}`
    : `${API_BASE}/orders/`
  const res = await fetch(url, {
    headers: { Authorization: `Bearer ${token}` },
  })
  if (!res.ok) {
    const data = await res.json().catch(() => ({}))
    throw new ApiError(res.status, data.detail ?? 'Failed to load orders')
  }
  const data = (await res.json()) as Record<string, unknown>[]
  return data.map(mapOrder)
}

export async function getOrder(token: string, orderId: string): Promise<Order> {
  const res = await fetch(`${API_BASE}/orders/${orderId}`, {
    headers: { Authorization: `Bearer ${token}` },
  })
  if (!res.ok) {
    const data = await res.json().catch(() => ({}))
    throw new ApiError(res.status, data.detail ?? 'Failed to load order')
  }
  return mapOrder(await res.json())
}

File: backend/app/api/v1/endpoints/order.py
from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.api.v1.endpoints.auth import get_current_user
from app.crud.cart_item import bulk_delete_cart_items
from app.crud.order import create_order, get_order_by_id, get_orders_by_user
from app.db.session import get_db
from app.models.order import Order, OrderStatus
from app.models.user import User
from app.schemas.order import (
    OrderCreate,
    OrderItemOut,
    OrderItemStoreMini,
    OrderOut,
)

router = APIRouter(prefix="/orders", tags=["orders"])


def _serialize(order: Order) -> OrderOut:
    items: list[OrderItemOut] = []
    for item in order.order_items:
        product = item.product
        store = product.store if product else None
        items.append(
            OrderItemOut(
                id=item.id,
                order_id=item.order_id,
                product_id=item.product_id,
                product_name=item.product_name,
                quantity=item.quantity,
                price=item.price,
                notes=item.notes,
                product_image=product.image if product else None,
                store=(
                    OrderItemStoreMini(
                        id=store.id,
                        name=store.name,
                        slug=store.slug,
                        avatar_url=store.avatar_url,
                    )
                    if store
                    else None
                ),
            )
        )
    return OrderOut(
        id=order.id,
        user_id=order.user_id,
        place=order.place,
        phone=order.phone,
        client_name=order.client_name,
        total_amount=order.total_amount,
        status=order.status,
        note=order.note,
        created_at=order.created_at,
        order_items=items,
    )


@router.post("/", response_model=OrderOut, status_code=status.HTTP_201_CREATED)
def create(
    body: OrderCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    if not body.items:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="Order must contain at least one item",
        )
    order = create_order(db, user_id=current_user.id, data=body)
    if body.cart_item_ids:
        bulk_delete_cart_items(db, user_id=current_user.id, ids=body.cart_item_ids)
    order = get_order_by_id(db, order.id)
    return _serialize(order)


@router.get("/", response_model=list[OrderOut])
def list_mine(
    status_filter: OrderStatus | None = Query(default=None, alias="status"),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    orders = get_orders_by_user(db, user_id=current_user.id, status=status_filter)
    return [_serialize(o) for o in orders]


@router.get("/{order_id}", response_model=OrderOut)
def get_one(
    order_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    order = get_order_by_id(db, order_id)
    if not order or order.user_id != current_user.id:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Order not found",
        )
    return _serialize(order)
